File: Projects-Solutions/Numbers/Go/pow.go
/* Fast Exponentiation:
 * Ask the user to enter 2 integers a and b and output a^b (i.e. pow(a,b)) in O(lg n) time complexity.
 */
package main

import "fmt"

/*pow: compute a**b using binary powering algorithm */
//TODO
//add comments on how this algorithm works
func pow(a, b int) int {
	p := 1
	for b > 0 {
		if b&1 != 0 {
			p *= a
		}
		b >>= 1
		a *= a
	}
	return p
}

//test the power function
func main() {
	var b, e int

	fmt.Println("Enter base")
	_, err := fmt.Scanf("%d", &b)

	if err != nil {
		panic(err)
	}

	fmt.Println("Enter exponential")
	_, err = fmt.Scanf("%d", &e)

	if err != nil {
		panic(err)
	}

	fmt.Println(pow(b, e))
}

File: Projects-Solutions/Numbers/Java/Pow.java
/* Fast Exponentiation:
 * Ask the user to enter 2 integers a and b and output a^b (i.e. pow(a,b)) in O(lg n) time complexity.
 */

import java.util.Scanner;

public class Pow {
	public static double power(double base, int exponential){
		double temp;
		//base case
		if(exponential == 0){
			return 1;
		}

		temp = power(base, exponential/2);

		//exponential is even
		if(exponential%2 == 0){
			return temp * temp;
		}else{
			//exponential is odd
			if(exponential > 0){
				//exponential is odd and positive
				return base * temp * temp;
			}else{
				//exponential is negative
				return (temp * temp)/base;
			}

		}
	}
	public static void main(String[] args){
		Scanner input = new Scanner(System.in);
		
		System.out.println("Enter base");
		int b = Integer.parseInt(input.nextLine());

		System.out.println("Enter exponential");
		int exp = Integer.parseInt(input.nextLine());

		System.out.println(power(b, exp));
	}
}
